walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,171 @@
+                //from startTime - the end of the current event time
+                if(currentEventTime.start() >= startTime + duration) {
+                    endTime = currentEventTime.start();
+                    break;
+                //if there is an overlap between the current event and the requested event, make the new startTime the end of the current event
+                } else if(currentEventTime.start() >= startTime && currentEventTime.start() < startTime + duration) {
+                    startTime = currentEventTime.end();
+                } 
+
+            }
+
+            //if the endTime of the time range (which was set to startTime) is greater than or equal to the duration of the meeting, it can be added as a new time range
+            if(endTime >= startTime + duration) {
+                availableTimes.add(TimeRange.fromStartEnd(startTime, endTime, false));
+            }
+
+            //takes care of overlapping meetings
+            //for example, currentEvent = meeting 2
+            //meeting 1: 8:30am - 10am
+            //meeting 2: 9am - 9:30am
+            //we want the startTime of the next timerange to still be 10am, not 9:30am because meeting 1 is still going on
+            startTime = currentEventTime.end() > startTime ? currentEventTime.end() : startTime;
+
+            //if we've looked at all meetings and there is enough time (until the end of the day) for the new meeting, create a new time range until the end of the day
+            if(i == sortedEventsList.size() && TimeRange.END_OF_DAY - startTime >= duration) {
+                availableTimes.add(TimeRange.fromStartEnd(startTime, TimeRange.END_OF_DAY, true));
+                break;
+            }
+
+        }
+
+        return availableTimes;
+
+    }
+
+    /**
+    * Creates a new ArrayList containing events with attendee overlap with the requested meeting. If no required attendees in the requested meeting
+    * are required at a specific event, that event will not be considered when creating TimeRange for the requested meeting.
+    * @return ArrayList<Event>
+    */
+    private ArrayList<Event> ignoreEventsWithoutRequestAttendees(Collection<Event> events, Collection<String> requestAttendees) {
+        ArrayList<Event> validEvents = new ArrayList<Event>();
+
+        for(Event event : events) {
+            if(anyAttendeeBusy(event, requestAttendees)) {
+                validEvents.add(event);
+            }
+        }
+
+        return validEvents;
+
+    }
+
+    /**
+    * Checks if there are any attendees required at the requested meeting that are also required at the specified Event.
+    * @return boolean, true if there is an attendee overlap, false otherwise
+    */
+    private boolean anyAttendeeBusy(Event event, Collection<String> attendees) {
+        Set<String> requestAttendees = new HashSet(attendees);
+        Set<String> eventAttendees = event.getAttendees();
+
+        for(String attendee : requestAttendees) {
```

Comment:
(optional): I think this should also work:

```
return requestAttendees
        .stream()
        .map(a -> eventAttendees.contains(a))
        .reduce(false, (a, b) -> a || b)
```

Reply:
Yep, just replaced the for loop with this.